Write to websocket asynchronously, plus fix thread-safety issues

Macil opened the pull request:
This fixes #3 and obsoletes #54. Additionally some thread-safety issues have been fixed so that all game-reading code happens from within the game thread. These two things have been done in separate commits. (I would have made them separate PRs except that my commit for the threading fixes depended on my async-send change.)

In https://github.com/opl-/beatsaber-http-status/issues/3#issuecomment-817007287, I theorized that the Send call inside of HTTPServer.StatusBroadcastBehavior.OnStatusChange was the cause of stuttering while using the mod. I tested this by adding some timing around the send call:

```csharp
			var s = eventJSON.ToString();

			var watch = new System.Diagnostics.Stopwatch();
			watch.Start();
			QueuedSend(s);
			watch.Stop();

			Console.WriteLine($"Execution Time: {watch.ElapsedMilliseconds} ms");
```

I got results like this:

<details><summary>Original Results</summary>

```
[INFO @ 07:42:20 | _] Execution Time: 48 ms
[INFO @ 07:42:26 | _] Execution Time: 0 ms
[INFO @ 07:42:26 | _] Execution Time: 0 ms
[INFO @ 07:42:27 | _] Execution Time: 0 ms
[INFO @ 07:42:27 | _] Execution Time: 0 ms
[INFO @ 07:42:27 | _] Execution Time: 0 ms
[INFO @ 07:42:27 | _] Execution Time: 0 ms
[INFO @ 07:42:27 | _] Execution Time: 0 ms
[INFO @ 07:42:27 | _] Execution Time: 0 ms
[INFO @ 07:42:27 | _] Execution Time: 49 ms
[INFO @ 07:42:27 | _] Execution Time: 0 ms
[INFO @ 07:42:27 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 0 ms
[INFO @ 07:42:28 | _] Execution Time: 46 ms
```
</details>

This is all within the main game thread, so the synchronous IO is regularly causing stuttering. (At 90fps, each frame has 11ms to happen, so blocking the main thread for an extra 40+ms regularly is pretty bad.) (I also did some timing around other parts of OnStatusChange and the rest of it was just zeroes.)

After the first commit, the results look perfect like this:

<details><summary>Fixed Results</summary>

```
[INFO @ 08:26:22 | _] Send execution Time: 0 ms
[INFO @ 08:26:22 | _] Send execution Time: 0 ms
[INFO @ 08:26:30 | _] Send execution Time: 0 ms
[INFO @ 08:26:30 | _] Send execution Time: 0 ms
[INFO @ 08:26:30 | _] Send execution Time: 0 ms
[INFO @ 08:26:30 | _] Send execution Time: 0 ms
[INFO @ 08:26:30 | _] Send execution Time: 0 ms
[INFO @ 08:26:31 | _] Send execution Time: 0 ms
[INFO @ 08:26:31 | _] Send execution Time: 0 ms
[INFO @ 08:26:31 | _] Send execution Time: 0 ms
[INFO @ 08:26:31 | _] Send execution Time: 0 ms
[INFO @ 08:26:31 | _] Send execution Time: 0 ms
[INFO @ 08:26:31 | _] Send execution Time: 0 ms
[INFO @ 08:26:31 | _] Send execution Time: 0 ms
[INFO @ 08:26:31 | _] Send execution Time: 0 ms
[INFO @ 08:26:31 | _] Send execution Time: 0 ms
[INFO @ 08:26:32 | _] Send execution Time: 0 ms
[INFO @ 08:26:32 | _] Send execution Time: 0 ms
[INFO @ 08:26:32 | _] Send execution Time: 0 ms
[INFO @ 08:26:32 | _] Send execution Time: 0 ms
[INFO @ 08:26:32 | _] Send execution Time: 0 ms
[INFO @ 08:26:32 | _] Send execution Time: 0 ms
[INFO @ 08:26:32 | _] Send execution Time: 0 ms
[INFO @ 08:26:32 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:33 | _] Send execution Time: 0 ms
[INFO @ 08:26:34 | _] Send execution Time: 0 ms
[INFO @ 08:26:34 | _] Send execution Time: 0 ms
[INFO @ 08:26:34 | _] Send execution Time: 0 ms
[INFO @ 08:26:34 | _] Send execution Time: 0 ms
[INFO @ 08:26:34 | _] Send execution Time: 0 ms
[INFO @ 08:26:34 | _] Send execution Time: 0 ms
[INFO @ 08:26:35 | _] Send execution Time: 0 ms
[INFO @ 08:26:35 | _] Send execution Time: 0 ms
[INFO @ 08:26:35 | _] Send execution Time: 0 ms
[INFO @ 08:26:35 | _] Send execution Time: 0 ms
[INFO @ 08:26:38 | _] Send execution Time: 0 ms
```
</details>

The asynchronous send fix properly queues up writes to the websocket.

Next, I noticed that HTTPServer would directly read from the StatusManager in response to web requests and websocket events even though StatusManager was written to from the main game thread. I made the code thread-safe by making sure that all access to StatusManager happened through the game thread.

(I did consider an alternate solution: It would be possible to make it so that StatusManager.EmitStatusUpdate wrote all of its results to a single volatile property, and then HTTPServer could read this volatile property when handling requests and websocket events, but this would have been a bigger change for negligible benefit that would make the code harder to evolve. In comparison, the change to just read from StatusManager within the game thread is a very small drop-in fix.)

opl- commented:
Huh, good job! I didn't expect a PR next day, but this looks *promising* at first glance. As I mentioned I currently don't have any way of testing this myself and I'm not very familiar with the C#/Unity threading APIs, so I'll have to figure out some way to run it myself and read up on some documentation. I'll report back as soon as I can.

Macil commented:
I've pushed a few more minor commits that make some minor enhancements and simplify some stuff. I read a bit more about how C# event properties work and turns out that changing the listeners is already thread-safe, so the parts where I added and removed listeners only through the main thread were unnecessary, as long as the event property was safely-checked before being called which I've down now.

opl- commented on BeatSaberHTTPStatus/HTTPServer.cs:
```diff
@@ -59,24 +63,38 @@ public class HTTPServer {
+			var promise = new TaskCompletionSource<object>();
+			var oldReadyToWrite = Interlocked.Exchange(ref readyToWrite, promise.Task);
+			oldReadyToWrite.ContinueWith(t => {
+				SendAsync(data, b => {
```
Considering this is already off-thread, is there any reason to use `SendAsync` and schedule yet another task?

Macil replied:
If there's no in-progress messages being sent at the time, the callback passed to `ContinueWith` will be executed within the main thread when it calls QueuedSend because of the `TaskContinuationOptions.ExecuteSynchronously` flag. (If there is a currently sending message, then the `ContinueWith` callback will later run on the websocket-sharp thread when `promise.SetResult(null)` is called.) Using the synchronous flag means the send can start immediately instead of only happening once a thread in the default thread pool gets around to it. No threads besides the game main thread and whatever websocket-sharp-managed threads are involved, and none of the threads have blocking IO happen on them.

If the `TaskContinuationOptions.ExecuteSynchronously` flag was not present as was the case in an earlier version of this PR, then it's a fair question. The code could instead look like this:

```csharp
		private void QueuedSend(string data) {
			var promise = new TaskCompletionSource<object>();
			var oldReadyToWrite = Interlocked.Exchange(ref readyToWrite, promise.Task);
			oldReadyToWrite.ContinueWith(t => {
				Send(data);
				promise.SetResult(null);
			}, connectionClosed.Token);
		}
```

This would still fix the issue of the game main thread being blocked, but it would have the downsides that the send wouldn't start until a thread from the thread pool was free to execute the callback and picked it up, and doing blocking IO on the default thread pool would tie up a thread for a period of time, possibly increasing the time it takes other thread pool tasks to start and maybe causing more threads to be instantiated. Using async IO would minimize our impact on the threadpool. (And then using the ExecuteSynchronously flag completely removes our impact and dependence on the threadpool.)

opl- replied:
I see, thanks for explaining!

opl- commented:
Also, I'd like to ask that you squash the commits before I merge.
